webviz_subsurface/plugins/_well_completions.py

```diff
@@ -0,0 +1,462 @@
+    open_frac = (
+        np.asarray(open_count_reduced, dtype=np.float64) / float(len(realizations))
+    ).round(decimals=3)
+
+    # calculate fraction of shut realizations
+    shut_count = (
+        np.minimum(np.asarray(compl_events), 0) * -1
+    )  # remove +1 and convert -1 to 1
+    shut_count_reduced = shut_count.sum(axis=0)  # sum over realizations
+    shut_frac = (
+        np.asarray(shut_count_reduced, dtype=np.float64) / float(len(realizations))
+    ).round(decimals=3)
+
+    # calculate khMean, khMin and khMax
+    kh_mean = (np.asarray(kh_values).sum(axis=0) / float(len(realizations))).round(
+        decimals=2
+    )
+    kh_min = np.asarray(kh_values).min(axis=0).round(decimals=2)
+    kh_max = np.asarray(kh_values).max(axis=0).round(decimals=2)
+
+    return open_frac, shut_frac, kh_mean, kh_min, kh_max
+
+
+def extract_well(df, well, zone_names, time_steps, realizations):
+    # pylint: disable=too-many-locals
+    """
+    Extract completion events and kh values for a single well
+    """
+    well_dict = {}
+    well_dict["name"] = well
+
+    compl_events, kh_values = get_completion_events_and_kh(df, zone_names, time_steps)
+    open_frac, shut_frac, kh_mean, kh_min, kh_max = calc_over_realizations(
+        compl_events, kh_values, realizations
+    )
+    result = {}
+    for (
+        zone_name,
+        open_frac_zone,
+        shut_frac_zone,
+        kh_mean_zone,
+        kh_min_zone,
+        kh_max_zone,
+    ) in zip(zone_names, open_frac, shut_frac, kh_mean, kh_min, kh_max):
+        result_zone = format_time_series(
+            open_frac_zone, shut_frac_zone, kh_mean_zone, kh_min_zone, kh_max_zone
+        )
+        if result_zone is not None:
+            result[zone_name] = result_zone
+    well_dict["completions"] = result
+    return well_dict
+
+
+def extract_wells(df, zone_names, time_steps, realizations, well_attributes):
+    """
+    Generates the wells part of the input dictionary to the WellCompletions component
+    """
+    well_list = []
+    for well_name, well_group in df.groupby("WELL"):
+        well_data = extract_well(
+            well_group, well_name, zone_names, time_steps, realizations
+        )
+        if not well_attributes is None and well_name in well_attributes:
+            well_data["attributes"] = well_attributes[well_name]
+        else:
+            well_data["attributes"] = {}
+        well_list.append(well_data)
+    return well_list
+
+
+def extract_stratigraphy(layer_zone_mapping, colors):
+    """
+    Returns the stratigraphy part of the data set
+    """
+    color_iterator = itertools.cycle(colors)
+    result = []
+    zones = []
+    for zone in layer_zone_mapping.values():
+        if zone not in zones:
+            zones.append(zone)
+            zdict = {}
+            zdict["name"] = zone
+            zdict["color"] = next(color_iterator)
+            result.append(zdict)
+    return result
```

Comment:
```suggestion
    return [{"name": zone, "color": next(color_iterator)} for zone in set(layer_zone_mapping.values)]
```
1) Is this equivalent (I haven't tested my suggested code above)?
2) The `zone not in zones` indicates that `layer_zone_mapping.values()` can't be assumed to contain unique values. Do you have a simple example illustrating what will duplicate zone name in the zone mapping?

Reply:
Yes, it is  because the layer_zone_mapping is on the format: {1:"Valysar", 2:"Valysar", ...}. But your solution was much more elegant, I didn't think about the set function. It should be equivalent, but I will test it.

Reply:
There is actually a problem with this solution and it is that it's not conserving the order of the zones. any suggestion for a solution to that?